pulser-core/pulser/math/__init__.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2024 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Custom implementation of math and array functions."""
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import cast, Protocol
+
+import numpy as np
+import scipy.fft
+
+from pulser.math.abstract_array import (
+    AbstractArray as AbstractArray,
+    AbstractArrayLike,
+)
+
+try:
+    import torch
+except ImportError:  # pragma: no cover
+    pass
+
+
+class Differentiable(Protocol):
+    """A type hint to signal that a parameter may be differentiable."""
+
+    # AbstractArray and torch.Tensor have this method, so it's a good proxy
+    def __array__(self) -> np.ndarray: ...
+
+
+# Custom function definitions
+
+
+def exp(a: AbstractArrayLike, /) -> AbstractArray:
+    a = AbstractArray(a)
+    if a.is_tensor:
+        return AbstractArray(torch.exp(a.as_tensor()))
+    return AbstractArray(np.exp(a.as_array()))
```

Comment:
I see a common pattern to all these functions... Perhaps we could have the pattern in a function `_func_abstr_array`

```suggestion
def _fun_abstr_array(a: AbstractArrayLike, torch_fun: Callable, np_fun:Callable) -> AbstractArray
    a = AbstractArray(a)
    if a.is_tensor:
        return AbstractArray(torch_fun(a.as_tensor()))
    return AbstractArray(np_fun(a.as_array()))

def exp(a: AbstractArrayLike, /) -> AbstractArray:
    return _fun_abstr_array(a, torch.exp, np.exp)

```

Reply:
As you like :) Perhaps some type errors might emerge with `Callable` indeed... I was also thinking about understandability, I am wondering if this solution does not decrease the understandability of the whole concept - even though I think it increases readability, since there are a lot of functions in these page...